src/FSharp.Data.GraphQL.Client/Subscriptions.fs

```diff
@@ -0,0 +1,247 @@
+            | "data" -> Data (getId r, getPayload r)
+            | "error" -> Error (getId r, getPayload r)
+            | "complete" -> Complete (getId r)
+            | other -> failwithf "Unexpected message type received from server. Can not treat \"%s\" message type." other
+        | other -> failwithf "Unexpected response JSON type. Expected a Record Json, but got %A." other
+
+/// A subscription response message, following the protocol of the handler that will use it.
+type IGraphQLSubscriptionResponse =
+    inherit IDisposable
+    /// When implemented, returns the immediate data for the operation.
+    abstract Data : JsonValue option
+    /// When implemented, returns the deferred data (@defer, @stream and subscriptions) for the operation.
+    abstract Deferred : IObservable<JsonValue>
+    /// When implemented, unsubscribe from the operation using the handler of the protocol.
+    abstract Unsubscribe : unit -> unit
+    /// When implemented, asynchronously unsubscribe from the operation using the handler of the protocol.
+    abstract AsyncUnsubscribe : unit -> Async<unit>
+
+type internal SubscriptionResponseMessage =
+    | Incoming of payload : JsonValue
+    | Finished
+
+/// The default GraphQL subscription response implementation that uses GraphQL Over Web Socket protocol.
+/// See https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md for information.
+type GraphQLSubscriptionResponse internal (source : IObservable<SubscriptionResponseMessage>, unsubscriber : unit -> Async<unit>) =
+    let tryPickFirstAndKeepRemaining (source : IObservable<'T>) =
+        let sourceSeq = Observable.toSeq source
+        let first = Seq.tryHead sourceSeq
+        match first with
+        | Some first -> Some (first, Seq.skip 1 sourceSeq)
+        | None -> None
+
+    let data, deferred =
+        source
+        |> tryPickFirstAndKeepRemaining
+        |> Option.map (fun (first, remaining) ->
+            let deferred = {
+                new IObservable<JsonValue> with
+                    member __.Subscribe(observer) =
+                        remaining
+                        |> Seq.iter (function
+                            | Incoming payload -> observer.OnNext(payload)
+                            | Finished -> observer.OnCompleted())
+                        { new IDisposable with member __.Dispose() = () } }
+            match first with
+            | Incoming payload -> Some payload, deferred
+            | Finished -> None, deferred)
+        |> Option.defaultValue (None, Observable.empty)
+
+    interface IGraphQLSubscriptionResponse with
+        member __.Data = data
+        member __.Deferred = deferred
+        member __.Unsubscribe() = unsubscriber () |> Async.RunSynchronously
+        member __.AsyncUnsubscribe() = unsubscriber ()
+
+    interface IDisposable with
+        member __.Dispose() = unsubscriber () |> Async.RunSynchronously
```

Comment:
Unsubscriber, in this case, will send a message to the server to end the connection (`Stop <id>`). When the server responds with `Complete <id>`, it will be processed in the observable above and call `OnComplete` to properly end the observer subscription.